File: paw-android/app/src/main/java/dev/paw/android/runtime/PawApiClient.kt
package dev.paw.android.runtime

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONObject
import java.io.BufferedReader
import java.io.InputStreamReader
import java.net.HttpURLConnection
import java.net.URL
import java.util.UUID

data class SendMessageResult(
    val id: String,
    val seq: Long,
    val createdAt: String,
)

class PawApiClient(
    private val baseUrl: String,
) {
    private var accessToken: String? = null

    fun setAccessToken(token: String?) {
        accessToken = token
    }

    suspend fun requestOtp(phone: String): JSONObject = requestJson(
        method = "POST",
        path = "/auth/request-otp",
        body = JSONObject().put("phone", phone),
    )

    suspend fun verifyOtp(phone: String, code: String): JSONObject = requestJson(
        method = "POST",
        path = "/auth/verify-otp",
        body = JSONObject().put("phone", phone).put("code", code),
    )

    suspend fun registerDevice(
        sessionToken: String,
        deviceName: String,
        ed25519PublicKey: String,
    ): JSONObject = requestJson(
        method = "POST",
        path = "/auth/register-device",
        body = JSONObject()
            .put("session_token", sessionToken)
            .put("device_name", deviceName)
            .put("ed25519_public_key", ed25519PublicKey),
    )

    suspend fun getMe(): JSONObject = requestJson(method = "GET", path = "/users/me")

    suspend fun updateMe(username: String, discoverableByPhone: Boolean): JSONObject = requestJson(
        method = "PATCH",
        path = "/users/me",
        body = JSONObject()
            .put("username", username)
            .put("discoverable_by_phone", discoverableByPhone),
    )

    suspend fun getConversations(): List<AndroidConversationItem> {
        val payload = requestJson(method = "GET", path = "/conversations")
        val rows = payload.optJSONArray("conversations") ?: JSONArray()
        return List(rows.length()) { index ->
            val row = rows.optJSONObject(index) ?: JSONObject()
            AndroidConversationItem(
                id = row.optString("id"),
                name = row.optString("name").ifBlank { "Conversation ${index + 1}" },
                lastMessage = row.optString("last_message").takeIf { it.isNotBlank() },
                unreadCount = row.optInt("unread_count", 0),
            )
        }
    }

    suspend fun getMessages(conversationId: String): List<AndroidChatMessage> {
        val payload = requestJson(method = "GET", path = "/conversations/$conversationId/messages")
        val rows = payload.optJSONArray("messages") ?: JSONArray()
        return List(rows.length()) { index ->
            val row = rows.optJSONObject(index) ?: JSONObject()
            AndroidChatMessage(
                id = row.optString("id").ifBlank { "msg-$index" },
                conversationId = row.optString("conversation_id").ifBlank { conversationId },
                senderId = row.optString("sender_id").ifBlank { "unknown" },
                content = row.optString("content"),
                format = row.optString("format").ifBlank { "plain" },
                seq = row.optLong("seq", 0L),
                createdAt = row.optString("created_at"),
                isMe = false,
                isAgent = false,
            )
        }.sortedBy { it.seq }
    }

    suspend fun sendMessage(
        conversationId: String,
        content: String,
    ): SendMessageResult {
        val payload = requestJson(
            method = "POST",
            path = "/conversations/$conversationId/messages",
            body = JSONObject()
                .put("content", content)
                .put("format", "plain")
                .put("idempotency_key", UUID.randomUUID().toString()),
        )
        return SendMessageResult(
            id = payload.optString("id"),
            seq = payload.optLong("seq", 0L),
            createdAt = payload.optString("created_at"),
        )
    }

    suspend fun registerPush(token: String) {
        requestJson(
            method = "POST",
            path = "/api/v1/push/register",
            body = JSONObject()
                .put("token", token)
                .put("platform", "fcm"),
        )
    }

    suspend fun unregisterPush() {
        requestJson(method = "DELETE", path = "/api/v1/push/register")
    }

    private suspend fun requestJson(
        method: String,
        path: String,
        body: JSONObject? = null,
    ): JSONObject = withContext(Dispatchers.IO) {
        val url = URL(baseUrl.trimEnd('/') + path)
        val connection = (url.openConnection() as HttpURLConnection).apply {
            requestMethod = method
            connectTimeout = 15_000
            readTimeout = 15_000
            doInput = true
            setRequestProperty("Content-Type", "application/json")
            accessToken?.takeIf { it.isNotBlank() }?.let {
                setRequestProperty("Authorization", "Bearer $it")
            }
            if (body != null) {
                doOutput = true
                outputStream.use { output ->
                    output.write(body.toString().toByteArray())
                }
            }
        }

        val status = connection.responseCode
        val stream = if (status in 200..299) connection.inputStream else connection.errorStream
        val payload = stream?.use {
            BufferedReader(InputStreamReader(it)).readText()
        }.orEmpty()
        connection.disconnect()

        if (status !in 200..299) {
            val message = runCatching {
                JSONObject(payload).optString("message").ifBlank {
                    JSONObject(payload).optString("error")
                }
            }.getOrDefault("HTTP $status")
            throw PawApiException(status, if (message.isBlank()) "HTTP $status" else message)
        }

        if (payload.isBlank()) {
            JSONObject()
        } else {
            JSONObject(payload)
        }
    }
}

class PawApiException(
    val statusCode: Int,
    override val message: String,
) : Exception(message)

File: paw-ios/Paw/Data/PawApiClient.swift
import Foundation

enum PawApiError: Error, LocalizedError {
    case invalidURL
    case httpError(statusCode: Int, message: String)
    case decodingError(String)
    case networkError(String)

    var errorDescription: String? {
        switch self {
        case .invalidURL:
            return "Invalid URL"
        case .httpError(let statusCode, let message):
            return "Server error (\(statusCode)): \(message)"
        case .decodingError(let detail):
            return "Failed to parse response: \(detail)"
        case .networkError(let detail):
            return "Network error: \(detail)"
        }
    }
}

actor PawApiClient {
    let baseURL: String
    private var accessToken: String?

    init(baseURL: String = PawApiClient.defaultBaseURL) {
        self.baseURL = baseURL
    }

    static var defaultBaseURL: String {
        #if targetEnvironment(simulator)
        "http://localhost:38173"
        #else
        "http://localhost:38173"
        #endif
    }

    func setAccessToken(_ token: String?) {
        accessToken = token
    }

    // MARK: - Auth

    func requestOtp(phone: String) async throws -> [String: Any] {
        try await post(path: "/auth/request-otp", body: ["phone": phone])
    }

    func verifyOtp(phone: String, code: String) async throws -> [String: Any] {
        try await post(path: "/auth/verify-otp", body: ["phone": phone, "code": code])
    }

    func registerDevice(
        sessionToken: String,
        deviceName: String,
        ed25519PublicKey: String
    ) async throws -> [String: Any] {
        try await post(path: "/auth/register-device", body: [
            "session_token": sessionToken,
            "device_name": deviceName,
            "ed25519_public_key": ed25519PublicKey,
        ])
    }

    func getMe() async throws -> [String: Any] {
        try await get(path: "/users/me")
    }

    func updateMe(username: String, discoverableByPhone: Bool) async throws -> [String: Any] {
        try await request(method: "PATCH", path: "/users/me", body: [
            "username": username,
            "discoverable_by_phone": discoverableByPhone,
        ])
    }

    // MARK: - Conversations

    func getConversations() async throws -> [[String: Any]] {
        let result = try await get(path: "/conversations")
        guard let conversations = result["conversations"] as? [[String: Any]] else {
            return []
        }
        return conversations
    }

    func getMessages(conversationId: String) async throws -> [[String: Any]] {
        let result = try await get(path: "/conversations/\(conversationId)/messages")
        guard let messages = result["messages"] as? [[String: Any]] else {
            return []
        }
        return messages
    }

    func sendMessage(
        conversationId: String,
        content: String
    ) async throws -> [String: Any] {
        try await post(
            path: "/conversations/\(conversationId)/messages",
            body: [
                "content": content,
                "format": "plain",
                "idempotency_key": UUID().uuidString,
            ]
        )
    }

    // MARK: - HTTP helpers

    private func get(path: String) async throws -> [String: Any] {
        try await request(method: "GET", path: path, body: nil)
    }

    private func post(path: String, body: [String: Any]) async throws -> [String: Any] {
        try await request(method: "POST", path: path, body: body)
    }

    private func request(
        method: String,
        path: String,
        body: [String: Any]?
    ) async throws -> [String: Any] {
        guard let url = URL(string: baseURL.trimmingSuffix("/") + path) else {
            throw PawApiError.invalidURL
        }

        var urlRequest = URLRequest(url: url)
        urlRequest.httpMethod = method
        urlRequest.setValue("application/json", forHTTPHeaderField: "Content-Type")
        urlRequest.timeoutInterval = 15

        if let token = accessToken, !token.isEmpty {
            urlRequest.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }

        if let body {
            urlRequest.httpBody = try JSONSerialization.data(withJSONObject: body)
        }

        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await URLSession.shared.data(for: urlRequest)
        } catch {
            throw PawApiError.networkError(error.localizedDescription)
        }

        guard let httpResponse = response as? HTTPURLResponse else {
            throw PawApiError.networkError("Invalid response type")
        }

        let statusCode = httpResponse.statusCode

        if data.isEmpty && statusCode >= 200 && statusCode < 300 {
            return [:]
        }

        guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            if statusCode >= 200 && statusCode < 300 {
                return [:]
            }
            throw PawApiError.decodingError("Non-JSON response (HTTP \(statusCode))")
        }

        guard statusCode >= 200 && statusCode < 300 else {
            let message = (json["message"] as? String)
                ?? (json["error"] as? String)
                ?? "HTTP \(statusCode)"
            throw PawApiError.httpError(statusCode: statusCode, message: message)
        }

        return json
    }
}

private extension String {
    func trimmingSuffix(_ suffix: String) -> String {
        if hasSuffix(suffix) {
            return String(dropLast(suffix.count))
        }
        return self
    }
}
